File: generate_colors.rb
require 'erb'
require 'json'

class ColorGenerator
  GENERATED_BLURB = "This is generated! Don't update manually!"

  RGB = Struct.new(:r, :g, :b) do
    def self.from_hex(val)
      chunks = val.scan(/[0-9a-f]{2}/i).map { |num| num.to_i(16) }
      self.new(*chunks)
    end

    def join(sep = ", ", &transform)
      transform ||= :itself
      [r, g, b].map(&transform).join(sep)
    end

    def join_intensity(sep = ", ")
      self.join(sep) { |item| (item / 255.0).round(2) }
    end

    def hex
      sprintf("#%02x%02x%02x", r, g, b)
    end
  end

  attr_reader :colors_hash

  def initialize(colors_json_path)
    # Load colors
    @colors_json = File.open(colors_json_path, "r")
    @colors_hash = JSON.load(@colors_json)
  end

  def call
    generate
    post_process
  end

  def generate
    safe_level = nil
    trim_mode = "%-"

    Dir.glob("tmpl/**/*") do |template_path|
      next if File.directory?(template_path)

      erb = ERB.new(IO.read(template_path), safe_level, trim_mode)
      target_path = template_path.sub("tmpl", "dist").chomp(".erb")
      IO.write(target_path, erb.result(binding))
    end
  end

  def post_process
    system "elm-format --yes dist/elm"
  end

  private

  def as_elm_name(raw)
    # prepend x, because x is cool
    #   e.g. 20 => x20
    raw.match?(/^[a-z]/) ? raw : "x" + raw
  end
end

def main
  generator = ColorGenerator.new("colors.json")
  generator.call
end

main

File: generate_colors.js
#!/usr/bin/env node

const fs = require("fs-extra");
const path = require("path");
const { exec } = require("child_process");

const Handlebars = require("handlebars");
const colorConvert = require("color-convert");
const klaw = require("klaw");

const COLORS_JSON = require("./colors.json");

/**
 * Convert the string into a usable programming identifier
 */
Handlebars.registerHelper("identifier", rawString => {
  // prepend x, because x is cool
  //   e.g. 20 => x20
  return /^[a-z]/.test(rawString) ? rawString : `x${rawString}`;
});


Handlebars.registerHelper("ternary", (predicate, yes, no) => predicate ? yes : no);
Handlebars.registerHelper("join", (sep, array) => array.join(sep));
Handlebars.registerHelper("keys", obj => Object.keys(obj));
Handlebars.registerHelper("rgb", hex => colorConvert.hex.rgb(hex));

function files(path) {
  return new Promise((resolve, reject) => {
    const paths = [];
    klaw(path)
      .on("data", item => item.stats.isFile() && paths.push(item.path))
      .on("error", reject)
      .on("end", () => resolve(paths));
  });
}

async function processTemplate(templatePath) {
  const source = await fs.readFile(templatePath, { encoding: "utf8" });
  const template = Handlebars.compile(source, { noEscape: true });
  const result = template({
    GENERATED_BLURB: "This is generated! Don't update manually!",
    colors: COLORS_JSON
  });

  const outPath = templatePath.replace("tmpl", "dist").replace(/.hbs$/, "");
  await fs.mkdirp(path.dirname(outPath));
  return fs.writeFile(outPath, result);
}

function postProcess() {
  return exec("elm-format --yes dist/elm");
}

async function main() {
  const paths = await files("./tmpl");
  await Promise.all(paths.map(processTemplate));
  return postProcess();
}

main();
